File: desktop/windows/control-panel/Services/RuntimeUiRequests.cs
namespace AgentDock.ControlPanel;

// UI-thread owned: writes invalidate reads, but hiding a window never releases an active mutation.
internal sealed class RuntimeUiRequests
{
    public long Revision { get; private set; }
    public bool Visible { get; private set; } = true;
    public bool Changing { get; private set; }
    private bool _reading;
    private long _change;
    public bool IsCurrent(long ticket) => Visible && ticket == Revision;
    public long? BeginRead() {
        if (!Visible || Changing || _reading) return null;
        _reading = true;
        return ++Revision;
    }
    public void EndRead(long ticket) { if (ticket == Revision) _reading = false; }
    public long? BeginChange() {
        if (!Visible || Changing) return null;
        Changing = true;
        _reading = false;
        return _change = ++Revision;
    }
    public void EndChange(long ticket) { if (ticket == _change) Changing = false; }
    public void SetVisible(bool visible) {
        Visible = visible;
        ++Revision;
        _reading = false;
    }
}

File: desktop/macos/AgentDockApp/Sources/RuntimeUIRequests.swift
// Owned by the UI thread. Closing invalidates responses without unlocking a running change.
final class RuntimeUIRequests {
    private(set) var revision = 0
    private(set) var visible = true
    private(set) var changing = false
    private var reading = false
    private var change = 0
    func isCurrent(_ ticket: Int) -> Bool { visible && ticket == revision }
    func beginRead() -> Int? {
        guard visible, !changing, !reading else { return nil }
        reading = true; revision += 1; return revision
    }
    func endRead(_ ticket: Int) { if ticket == revision { reading = false } }
    func beginChange() -> Int? {
        guard visible, !changing else { return nil }
        changing = true; reading = false; revision += 1; change = revision; return change
    }
    func endChange(_ ticket: Int) { if ticket == change { changing = false } }
    func setVisible(_ value: Bool) { visible = value; revision += 1; reading = false }
}
